### PR TITLE
feat(jellyfish-api-core): Add account getPendingFutureSwaps RPC for dfip2203 futures

### LINKED ISSUE #1268: `packages/jellyfish-api-core`: Add DFI2203 Futures work
#### What would you like to be added:
map endpoints related to https://github.com/DeFiCh/ain/pull/1155

#### Why is this needed:
https://github.com/DeFiCh/dfips/issues/127

### PR DESCRIPTION
**Pull request by jingyi2811**
#### What this PR does / why we need it:
Add getPendingFutureSwaps RPC

#### Which issue(s) does this PR fixes?:

Fixes part of #1268

### CODE REVIEW

**Comment by jingyi2811** [comment]
The final review of this PR is going to determine how the design of listPendingFutureSwaps PR looks like as both PRs are highly related.

File: packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+        const pendingFutureSwaps = await testing.rpc.account.getPendingFutureSwaps(tslaAddress)
+        expect(pendingFutureSwaps).toStrictEqual({
+          owner: tslaAddress,
+          values: []
+        })
+      }
+    })
+
+    it('Should getPendingFutureSwaps if futureswap DUSD for TSLA', async () => {
+      // Call getpendingfutureswaps before performing futureswap
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts:148]: It looks redundant as Line 139 does the same thing.
However, I believe every code in every Jest Item (It) should be independent.

File: packages/testcontainers/src/containers/DeFiDContainer.ts [hunk 2]
```diff
@@ -35,7 +35,7 @@ export abstract class DeFiDContainer extends DockerContainer {
     if (process?.env?.DEFICHAIN_DOCKER_IMAGE !== undefined) {
       return process.env.DEFICHAIN_DOCKER_IMAGE
     }
-    return 'defi/defichain:master-2a236bb79'
+    return 'defi/defichain:master-57c099c1a'
```
Comment (jingyi2811) [inline packages/testcontainers/src/containers/DeFiDContainer.ts:38]: This should be the latest docker, if not reviewer please change this for me.

File: packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+      symbol: 'DUSD',
+      amount: 4
+    })
+    await testing.generate(1)
+
+    await testing.token.mint({
+      symbol: 'TSLA',
+      amount: 4
+    })
+    await testing.generate(1)
```
Comment (canonbrother) [inline packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts:65]: need change the loan token source from "mint" to "takeLoan from vault"

File: packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+      await testing.generate(1)
+
+      await testing.container.call('futureswap', [tslaAddress, '0.6@TSLA'])
+      await testing.generate(1)
+
+      await testing.container.call('futureswap', [tslaAddress, '0.5@TSLA'])
+      await testing.generate(1)
+
+      await testing.container.call('futureswap', [tslaAddress, '0.2@TSLA'])
+      await testing.generate(1)
```
Comment (canonbrother) [inline packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts:220]: one more test request - check the `getPendingFutureSwaps` result on same block testing
```
futureswap(1)
futureswap(2)
gen(1)
```

**Review by canonbrother — COMMENTED** [review]
check the listing logic against the gov
```
futureswap
check the list 
deactivate the DFIP2203 
check the list again
```

File: packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts (hunk 3 above)
Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts:65]: What is the reason? 

Peter does not do so in https://github.com/DeFiCh/ain/blob/master/test/functional/feature_futures.py

Maybe I miss the context as I had been taking long leave for some time

Comment (canonbrother) [reply packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts:65]: https://github.com/DeFiCh/jellyfish/issues/865

File: packages/jellyfish-api-core/src/category/account.ts [hunk 5]
```diff
@@ -552,3 +561,13 @@ export interface BurnInfo {
...
   paybacktokens: string[]
 }
+
+export interface GetFutureInfo {
+  owner: string
+  values: FutureData[]
+}
+
+export interface FutureData {
+  source: string
```
Comment (ivan-zynesis) [inline packages/jellyfish-api-core/src/category/account.ts:571]:
```suggestion
  source: string // eg: '1.234@DUSD'
```
Am thinking should we stick to ain's naming too. (Although I think it is still confusing)
https://github.com/DeFiCh/ain/pull/1155/files#diff-84c7a7f36362b9724c31e5dec9879b2f81eae0d0addbc9c0933c3558c577de65R218-R221

File: packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+        'v0/params/dfip2203/block_period': futureInterval.toString()
+      }
+    })
+    await testing.generate(1)
+
+    await testing.rpc.masternode.setGov({ ATTRIBUTES: { 'v0/params/dfip2203/active': 'true' } })
+    await testing.generate(1)
+  }
+
+  beforeEach(async () => {
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts:141]: Reset container to remove attributes

File: packages/jellyfish-testing/src/token.ts [hunk 7]
```diff
@@ -43,6 +43,11 @@ export class TestingToken {
     const to = { [address]: [account] }
     return await this.rpc.account.sendTokensToAddress({}, to)
   }
+
+  async getTokenId (symbol: string): Promise<string> {
```
Comment (jingyi2811) [inline packages/jellyfish-testing/src/token.ts:47]: This is a new function created by Suranga

File: packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts [hunk 8]
```diff
@@ -0,0 +1,598 @@
...
+          expect(pendingFutureSwaps).toStrictEqual({
+            owner: tslaAddress,
+            values: []
+          })
+        }
+      })
+    })
+  })
+
+  describe('Multiple futureswaps', () => {
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/account/getPendingFutureSwaps.test.ts:449]: I don't duplicate the negative test units for setGov in **multiple** futureswaps

**Review by surangap — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**